### PR TITLE
18 Implement GameState serialization and deserialization

### PR DESCRIPTION
**Pull request by Artewald**

### CODE REVIEW

**Comment by Artewald** [comment]
To keep on working I had to merge the ECS branch into this one. If we don't want to merge it into the main branch, we could possibly merge into ECS instead.

File: core/src/com/softwarearchitecture/ECS/Card.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.softwarearchitecture.ecs;
+
+public interface Card {
```
Comment (SverreNystad) [inline core/src/com/softwarearchitecture/ECS/Card.java:3]: This is a very important concept in our application and as it is an empty interface it should have javadocs to describe what this is

File: core/src/com/softwarearchitecture/ECS/ComponentManager.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+     * @param component The component to add to the entity.
+     */
+    public void addComponent(Entity entity, T component) {
+        components.put(entity.getId(), component);
+    }
+
+    /**
+     * Retrieves a component associated with a given entity. If the entity does not
+     * have
+     * a component of this type, {@code null} is returned.
```
Comment (SverreNystad) [inline core/src/com/softwarearchitecture/ECS/ComponentManager.java:32]: We should use optional rather then Null

**Comment by SverreNystad** [comment]
While reviewing I noticed that I came with review more toward the ECS branch than this branch. To not be redundant I will focus on the parts that are specific to this branch

File: core/src/com/softwarearchitecture/game_server/MapTile.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.softwarearchitecture.game_server;
+
+public enum MapTile {
```
Comment (SverreNystad) [inline core/src/com/softwarearchitecture/game_server/MapTile.java:3]: Maybe add javadocs

**Comment by SverreNystad** [comment]
lgtmhrtc!

**Pull request merged**